vault/rollback.go

```diff
@@ -216,23 +247,25 @@ func (m *RollbackManager) attemptRollback(ctx context.Context, fullPath string,
 
 // Rollback is used to trigger an immediate rollback of the path,
 // or to join an existing rollback operation if in flight. Caller should have
-// core's statelock held
+// core's statelock held (write OR read). If an already inflight rollback is
+// happening this function will simply wait for it to complete
 func (m *RollbackManager) Rollback(ctx context.Context, path string) error {
 	ns, err := namespace.FromContext(ctx)
 	if err != nil {
 		return err
 	}
 	fullPath := ns.Path + path
 
-	// Check for an existing attempt and start one if none
-	m.inflightLock.RLock()
-	rs, ok := m.inflight[fullPath]
-	m.inflightLock.RUnlock()
-	if !ok {
-		rs = m.startRollback(ctx, fullPath, false)
-	}
+	// Check for an existing attempt or start one if none
+	rs := m.startOrLookupRollback(ctx, fullPath, false)
+
+	// Since we have the statelock held, tell any inflight rollback to give up
+	// trying to aquire it. This will prevent deadlocks in the case where we
+	// have the write lock.
```

Comment:
```suggestion
	// have the write lock. In the case where it was waiting to grab
	// a read lock it will then simply continue with the rollback
	// operation under the protection of our write lock.
```